dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModelManager.java

```diff
@@ -0,0 +1,669 @@
+
+        relModelCache.clear();
+        relGraphCache.clear();
+        edgeModelCache.clear();
+        extractRelations(relsToExtract);
+
+        this.context = null;
+        this.model = null;
+        this.wmm = null;
+        this.domain = null;
+
+        return executionModel;
+    }
+
+    private void extractEvents() {
+        int id = 0;
+        List<Thread> threadList = new ArrayList<>(context.getTask().getProgram().getThreads());
+
+        for (Thread t : threadList) {
+            ThreadModel tm = new ThreadModel(t);
+            int eventNum = 0;
+            Event e = t.getEntry();
+
+            do {
+                if (!isTrue(context.execution(e))) {
+                    e = e.getSuccessor();
+                    continue;
+                }
+
+                if (toExtract(e)) {
+                    extractEvent(e, tm, id++);
+                    eventNum++;
+                }
+                
+                if (e instanceof CondJump jump
+                    && isTrue(context.jumpCondition(jump))) {
+                    e = jump.getLabel();
+                } else {
+                    e = e.getSuccessor();
+                }
+
+            } while (e != null);
+
+            if (eventNum > 0) { executionModel.addThread(tm); }
+        }
+    }
+
+    private void extractEvent(Event e, ThreadModel tm, int id) {
+        EventModel em;
+        if (e instanceof MemoryEvent memEvent) {
+            Object addressObj = checkNotNull(
+                evaluateByModel(context.address(memEvent))
+            );
+            final BigInteger address = new BigInteger(addressObj.toString());
+            
+            String valueString = String.valueOf(
+                evaluateByModel(context.value(memEvent))
+            );
+            final BigInteger value = switch(valueString) {
+                // NULL case can happen if the solver optimized away a variable.
+                // This should only happen if the value is irrelevant, so we will just pick 0.
+                case "false", "null" -> BigInteger.ZERO;
+                case "true" -> BigInteger.ONE;
+                default -> new BigInteger(valueString);
+            };
+
+            if (memEvent.hasTag(Tag.READ)) {
+                em = new LoadModel(memEvent, tm, id, address, value);
+                executionModel.addAddressRead(address, (LoadModel) em);
+            } else if (memEvent.hasTag(Tag.WRITE)) {
+                em = new StoreModel(memEvent, tm, id, address, value);
+                executionModel.addAddressWrite(address, (StoreModel) em);
+            } else {
+                // Should never happen.
+                throw new IllegalArgumentException(String.format(
+                    "Event %s is memory event but neither read nor write", memEvent
+                ));
+            }
+
+        } else if (e.hasTag(Tag.FENCE) && e instanceof GenericVisibleEvent visible) {
+            final String name = (visible).getName();
```

Comment:
I think the fence check is not helpful. All visible events should be extracted, no?

Reply:
do you mean `e.hasTag(Tag.FENCE)`? This if clause is supposed to create a fence model. While I think so far, all our `GenericVisibleEvent` are indeed fences, I don't think we should rely on this assumption.

Reply:
I don't think all of them are fences. IIRC, some GPU events are not. Either way, making the distinction of fence vs non-fence doesn't do much. `FenceModel` might as well be a `GenericVisibleEventModel`.

Reply:
> FenceModel might as well be a GenericVisibleEventModel.

This is indeed the best solution